chain/events/filter/index.go

```diff
@@ -520,6 +640,61 @@ func (ei *EventIndex) Close() error {
 	return ei.db.Close()
 }
 
+func (ei *EventIndex) SubscribeUpdates() (uint64, chan *EventIndexUpdate) {
+	subCtx, subCancel := context.WithCancel(context.Background())
+	ch := make(chan *EventIndexUpdate)
+
+	tSub := &updateSub{
+		ctx:    subCtx,
+		cancel: subCancel,
+		ch:     ch,
+	}
+
+	ei.mu.Lock()
+	subId := ei.subIdCounter
+	ei.subIdCounter++
+	ei.updateSubs[subId] = tSub
+	ei.mu.Unlock()
+
+	return subId, tSub.ch
+}
+
+func (ei *EventIndex) UnsubscribeUpdates(subId uint64) {
+	ei.mu.Lock()
+	tSub, ok := ei.updateSubs[subId]
+	if !ok {
+		ei.mu.Unlock()
+		return
+	}
+	delete(ei.updateSubs, subId)
+	ei.mu.Unlock()
+
+	// cancel the subscription
+	tSub.cancel()
+}
+
+func (ei *EventIndex) GetMaxHeightInIndex(ctx context.Context) (uint64, error) {
+	row := ei.db.QueryRowContext(ctx, "SELECT MAX(height) FROM events_seen")
+	var maxHeight uint64
+	err := row.Scan(&maxHeight)
+	return maxHeight, err
+}
+
+func (ei *EventIndex) IsHeightProcessed(ctx context.Context, height uint64) (bool, error) {
+	mh, err := ei.GetMaxHeightInIndex(ctx)
+	if err != nil {
+		return false, err
+	}
+	return height <= mh, nil
+}
+
+func (ei *EventIndex) IsTipsetProcessed(ctx context.Context, tipsetKeyCid []byte) (bool, error) {
+	row := ei.db.QueryRowContext(ctx, "SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key_cid = ?", tipsetKeyCid)
```

Comment:
this should probably be a prepared statement

Reply:
Done.